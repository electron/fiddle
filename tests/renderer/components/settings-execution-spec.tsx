import * as React from 'react';

import { shallow } from 'enzyme';

import {
  ExecutionSettings,
  SettingItemType,
} from '../../../src/renderer/components/settings-execution';

describe('ExecutionSettings component', () => {
  let store: any;

  beforeEach(() => {
    store = {
      executionFlags: [],
      environmentVariables: [],
    };
  });

  it('renders', () => {
    const wrapper = shallow(<ExecutionSettings appState={store} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleDeleteDataChange()', () => {
    it('handles a new selection', async () => {
      const wrapper = shallow(<ExecutionSettings appState={store} />);
      const instance = wrapper.instance() as any;
      await instance.handleDeleteDataChange({
        currentTarget: { checked: false },
      });

      expect(store.isKeepingUserDataDirs).toBe(false);

      await instance.handleDeleteDataChange({
        currentTarget: { checked: true },
      });

      expect(store.isKeepingUserDataDirs).toBe(true);
    });
  });

  describe('handleElectronLoggingChange()', () => {
    it('handles a new selection', async () => {
      const wrapper = shallow(<ExecutionSettings appState={store} />);
      const instance = wrapper.instance() as any;
      await instance.handleElectronLoggingChange({
        currentTarget: { checked: false },
      });

      expect(store.isEnablingElectronLogging).toBe(false);

      await instance.handleElectronLoggingChange({
        currentTarget: { checked: true },
      });

      expect(store.isEnablingElectronLogging).toBe(true);
    });
  });

  describe('handleItemSettingsChange()', () => {
    describe('with executionFlags', () => {
      it('updates when new flags are added', async () => {
        const wrapper = shallow(<ExecutionSettings appState={store} />);
        const instance = wrapper.instance() as any;

        const lang = '--lang=es';
        const flags = '--js-flags=--expose-gc';

        await instance.handleSettingsItemChange(
          {
            currentTarget: { name: '0', value: lang },
          },
          SettingItemType.Flags,
        );

        expect(instance.state.executionFlags).toEqual({ '0': lang });
        expect(store.executionFlags).toEqual([lang]);

        await instance.handleSettingsItemChange(
          {
            currentTarget: { name: '1', value: flags },
          },
          SettingItemType.Flags,
        );

        expect(instance.state.executionFlags).toEqual({
          '0': lang,
          '1': flags,
        });
        expect(store.executionFlags).toEqual([lang, flags]);
      });
    });

    describe('with environmentVariables', () => {
      it('updates when new flags are added', async () => {
        const wrapper = shallow(<ExecutionSettings appState={store} />);
        const instance = wrapper.instance() as any;

        const debug = 'ELECTRON_DEBUG_DRAG_REGIONS=1';
        const trash = 'ELECTRON_TRASH=trash-cli';

        await instance.handleSettingsItemChange(
          {
            currentTarget: {
              name: '0',
              value: debug,
            },
          },
          SettingItemType.EnvVars,
        );

        expect(instance.state.environmentVariables).toEqual({ '0': debug });
        expect(store.environmentVariables).toEqual([debug]);

        await instance.handleSettingsItemChange(
          {
            currentTarget: { name: '1', value: trash },
          },
          SettingItemType.EnvVars,
        );

        expect(instance.state.environmentVariables).toEqual({
          '0': debug,
          '1': trash,
        });
        expect(store.environmentVariables).toEqual([debug, trash]);
      });
    });
  });
});
