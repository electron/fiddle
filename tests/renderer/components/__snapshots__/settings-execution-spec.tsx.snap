// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecutionSettings component renders 1`] = `
<div>
  <h1>
    Execution
  </h1>
  <Blueprint3.Callout>
    These advanced settings control how Electron Fiddle executes your fiddles.
  </Blueprint3.Callout>
  <br />
  <Blueprint3.Callout>
    <Blueprint3.FormGroup>
      <p>
        Whenever Electron runs, it creates a user data directory for cookies, the cache, and various other things that it needs to keep around. Since fiddles are usually just run once, we delete this directory after your fiddle exits. Enable this setting to keep the user data directories around.
      </p>
      <Blueprint3.Checkbox
        label="Do not delete user data directories."
        onChange={[Function]}
      />
    </Blueprint3.FormGroup>
  </Blueprint3.Callout>
  <br />
  <Blueprint3.Callout>
    <Blueprint3.FormGroup>
      <p>
        There are some flags that Electron uses to log extra information both internally and through Chromium. Enable this option to make Fiddle produce those logs. Enabling advanced Electron logging will set the 
        <code>
          ELECTRON_ENABLE_LOGGING
        </code>
        ,
         
        <code>
          ELECTRON_DEBUG_NOTIFICATION
        </code>
        , and
         
        <code>
          ELECTRON_ENABLE_STACK_DUMPING
        </code>
         environment variables to true. See
         
        <a
          href="https://www.electronjs.org/docs/api/environment-variables"
        >
          documentation
        </a>
         
        for more information about what they do.
      </p>
      <Blueprint3.Checkbox
        label="Enable advanced Electron logging."
        onChange={[Function]}
      />
    </Blueprint3.FormGroup>
  </Blueprint3.Callout>
  <br />
  <Blueprint3.Callout>
    <Blueprint3.FormGroup>
      <p>
        Electron allows starting the executable with
         
        <a
          href="https://www.electronjs.org/docs/api/command-line-switches"
        >
          user-provided flags
        </a>
        , such as 
        <code>
          --js-flags=--expose-gc
        </code>
        . Those can be added to run when you start your Fiddles.
      </p>
      <br />
      <Blueprint3.InputGroup
        aria-label="Set user-provided flags"
        key="0"
        name="0"
        onChange={[Function]}
        placeholder="--js-flags=--expose-gc"
        rightElement={
          <Blueprint3.Button
            disabled={true}
            icon="cross"
            onClick={[Function]}
          />
        }
        value=""
      />
    </Blueprint3.FormGroup>
    <Blueprint3.ButtonGroup>
      <Blueprint3.Button
        onClick={[Function]}
      >
        Add New Flag
      </Blueprint3.Button>
    </Blueprint3.ButtonGroup>
  </Blueprint3.Callout>
  <br />
  <Blueprint3.Callout>
    <Blueprint3.FormGroup>
      <p>
        Electron allows starting the executable with
         
        <a
          href="https://www.electronjs.org/docs/api/environment-variables"
        >
          user-provided environment variables
        </a>
        , such as
         
        <code>
          NODE_OPTIONS="--no-warnings --max-old-space-size=2048"
        </code>
        . Those can be added here to run when you start your Fiddles.
      </p>
      <br />
      <Blueprint3.InputGroup
        aria-label="Set user-provided environment variables"
        key="0"
        name="0"
        onChange={[Function]}
        placeholder="NODE_OPTIONS=\\"--no-warnings --max-old-space-size=2048\\""
        rightElement={
          <Blueprint3.Button
            disabled={true}
            icon="cross"
            onClick={[Function]}
          />
        }
        value=""
      />
    </Blueprint3.FormGroup>
    <Blueprint3.ButtonGroup>
      <Blueprint3.Button
        onClick={[Function]}
      >
        Add New Variable
      </Blueprint3.Button>
    </Blueprint3.ButtonGroup>
  </Blueprint3.Callout>
  <br />
  <Blueprint3.Callout>
    <Blueprint3.FormGroup>
      <span
        style={
          Object {
            "marginRight": 4,
          }
        }
      >
        Electron Fiddle will install packages if you specify them. It uses
         
        <a
          href="https://www.npmjs.com/"
          rel="noreferrer"
          target="_blank"
        >
          npm
        </a>
         
        as its package manager by default, but
         
        <a
          href="https://classic.yarnpkg.com/lang/en/"
          rel="noreferrer"
          target="_blank"
        >
          Yarn
        </a>
         
        is also available.
      </span>
      <Blueprint3.RadioGroup
        inline={true}
        onChange={[Function]}
      >
        <Blueprint3.Radio
          label="npm"
          value="npm"
        />
        <Blueprint3.Radio
          label="yarn"
          value="yarn"
        />
      </Blueprint3.RadioGroup>
    </Blueprint3.FormGroup>
  </Blueprint3.Callout>
</div>
`;
